SAI/sai_api_gen.py

```diff
@@ -0,0 +1,241 @@
+    elif key_size <= 16:
+        return 'sai_u16_range_list_t'
+    elif key_size == 32 and ('addr' in key_field or 'ip' in key_header):
+        return 'sai_ipaddr_range_list_t'
+    elif key_size <= 32:
+        return 'sai_u32_range_list_t'
+    elif key_size <= 64:
+        return 'sai_u64_range_list_t'
+
+
+def get_sai_key_data(program, key):
+    sai_key_data = dict()
+    full_key_name, sai_key_name = key[NAME_TAG].split(':')
+    key_tuple = full_key_name.split('.')
+    if len(key_tuple) == 3:
+        key_struct, key_header, key_field = key_tuple
+    else:
+        key_header, key_field = key_tuple
+    sai_key_data['sai_key_name'] = sai_key_name
+
+    key_header_type = None
+    for header in program['headers']:
+        if header['name'] == key_header:
+            key_header_type = header['header_type']
+
+    key_size = 0
+    for header_type in program['header_types']:
+        if header_type['name'] == key_header_type:
+            for field in header_type['fields']:
+                if field[0] == key_field:
+                    key_size = int(field[1])
+
+    sai_key_data['match_type'] =  key['match_type']
+    if sai_key_data['match_type'] == 'exact':
+        sai_key_data['sai_key_type'] = get_sai_key_type(key_size, key_header, key_field)
+    elif sai_key_data['match_type'] == 'lpm':
+        sai_key_data['sai_lpm_type'] = get_sai_lpm_type(key_size, key_header, key_field)
+    elif sai_key_data['match_type'] == 'list':
+        sai_key_data['sai_list_type'] = get_sai_list_type(key_size, key_header, key_field)
+    elif sai_key_data['match_type'] == 'range_list':
+        sai_key_data['sai_range_list_type'] = get_sai_range_list_type(key_size, key_header, key_field)
+
+    return sai_key_data
+
+
+def get_sai_action_data(program, action_name):
+    sai_action_data = dict()
+    sai_action_data['name'] = action_name.split('.')[-1]
+    params = []
+
+    for action in program['actions']:
+        if action['name'] == action_name:
+            for rdata in action['runtime_data']:
+                param = dict()
+                param['name'] = rdata['name']
+                param['type'] = get_sai_key_type(int(rdata['bitwidth']), param['name'], param['name'])
+                params.append(param)
+
+    sai_action_data['params'] = params
+    return sai_action_data
+
+
+def generate_sai_api(program, ignore_tables):
+    sai_api = dict()
+    tables = get_tables(get_ingress_pipeline(program))
+    sai_tables = []
+    for table in tables:
+        sai_table_data = dict()
+        sai_table_data['keys'] = []
+        sai_table_data['actions'] = []
+        table_control, table_name = table[NAME_TAG].split('.', 1)
+        sai_table_data['name'] = table_name.replace('.' , '_')
+
+        if sai_table_data['name'] in ignore_tables:
+            continue
+
+        for key in table[KEY_TAG]:
+            sai_table_data['keys'].append(get_sai_key_data(program, key))
+
+        for action in table['actions']:
```

Comment:
Can you extract the P4 default action and render a suitable comment in the generated header file? For example, ACL default actions are `deny` and this should appear in the generated .h. Likewise, if there are other attributes or subtleties which are buried or implied by the P4 code, can we try to make them explicit in the generated .h header? If you cannot add to the current PR, perhaps create an issue for this so we can track it. Thanks.

Reply:
I believe it is possible, but it is outside of the scope of this commit. Can you open an issue?

Reply:
Done, see https://github.com/Azure/DASH/issues/47